java-checks/src/main/java/org/sonar/java/checks/spring/NullableInjectedFieldsHaveDefaultValueCheck.java

```diff
@@ -0,0 +1,150 @@
+    return extractLiteralTree(annotation.arguments().get(0)).map(literalTree -> {
+      // Compute replacement
+      String currentValue = literalTree.value();
+      String replacementValue = currentValue.substring(0, currentValue.lastIndexOf('}'))
+        + ":#{null}}"
+        + "\"";
+      // Insert replacement
+      return List.of(
+        JavaQuickFix.newQuickFix("Set null as default value")
+          .addTextEdit(JavaTextEdit.replaceTree(literalTree.token(), replacementValue))
+          .build()
+      );
+    }).orElse(List.of());
+  }
+
+  /**
+   * Maps a class member to a {@link AnnotationsOfInterest} if it is a field annotated as nullable and injected with a Spring value annotation.
+   *
+   * @param member A class member (eg: method, field, ...)
+   * @return An Optional with a {@link AnnotationsOfInterest} if the member matches. Optional.empty() otherwise.
+   */
+  private static Optional<AnnotationsOfInterest> mapToTreesOfInterest(Tree member) {
+    if (!member.is(Tree.Kind.VARIABLE)) {
+      return Optional.empty();
+    }
+    var field = (VariableTree) member;
+    Optional<AnnotationTree> valueAnnotation = getValueAnnotationWithoutDefault(field);
+    if (valueAnnotation.isEmpty()) {
+      return Optional.empty();
+    }
+    Optional<AnnotationTree> nullableAnnotation = getNullableAnnotation(field);
+    return nullableAnnotation.map(annotationTree -> new AnnotationsOfInterest(valueAnnotation.get(), annotationTree));
+  }
+
+  private static Optional<AnnotationTree> getValueAnnotationWithoutDefault(VariableTree field) {
+    return field.modifiers().annotations().stream()
+      .filter(annotation -> annotation.symbolType().is(VALUE_ANNOTATION) && !hasDefaultValue(annotation))
+      .findFirst();
+  }
+
+  private static Optional <AnnotationTree> getNullableAnnotation(VariableTree field) {
+    SymbolMetadata.NullabilityData nullabilityData = field.symbol().metadata().nullabilityData(SymbolMetadata.NullabilityTarget.FIELD);
+    SymbolMetadata.AnnotationInstance instance = nullabilityData.annotation();
+    if (instance == null) {
+      return Optional.empty();
+    }
+    return Optional.ofNullable(field.symbol().metadata().findAnnotationTree(instance));
+  }
+
+  private static boolean hasDefaultValue(AnnotationTree annotation) {
+    ExpressionTree expression = annotation.arguments().get(0);
+    return extractLiteralValue(expression)
+      .map(argument -> argument.contains(":"))
+      .orElse(true);
+  }
+
+  private static Optional<String> extractLiteralValue(ExpressionTree annotationArgument) {
+    return extractLiteralTree(annotationArgument)
+      .flatMap(literal -> literal.asConstant(String.class));
+  }
+
+  private static Optional<LiteralTree> extractLiteralTree(ExpressionTree annotationArgument) {
+    if (annotationArgument.is(Tree.Kind.STRING_LITERAL)) {
+      return Optional.of((LiteralTree) annotationArgument);
+    }
+    // If the argument is not a literal, then it must be an assignment
```

Comment:
It could also be a variable (the `WEIRD` case you have in the sample file). And we have some logic already to deal with that - you can use `ExpressionHelper#getConstantValueAsString`. If it isn't an assignment you can pass the argument directly to that method, otherwise the assignment value.

Reply:
Hmm when I set a breakpoint on line 159 when visiting sample line 41, I can call:
```
ExpressionsHelper.getConstantValueAsString(expression).value()
```
And it will return `${my.constant:#{null}}`.
I think the problem here may be that the line 41 should read:
```
  @Value(value = NON_COMPLIANT_IN_A_CONSTANT) // Compliant FN
```
but currently reads:
```
  @Value(value = DEFINED_IN_A_CONSTANT) // Compliant FN
```

In any case, the other problem remains that the annotation argument could be a variable (this case is currently missing in the sample file). If I add e.g. `@Value(DEFINED_IN_A_CONSTANT)` somewhere, it will cause a `ClassCastException` on line 155.